src/Pos/Worker.hs

```diff
@@ -58,6 +60,12 @@ allWorkers NodeResources {..} = mconcatPair
     , wrap' "delegation" $ dlgWorkers
     , wrap' "slotting"   $ (properSlottingWorkers, mempty)
 
+    , case ncTopology ncNetworkConfig of
+        TopologyBehindNAT dnsDomains ->
+          wrap' "subscription" $ subscriptionWorkers ncNetworkConfig dnsDomains
+        _otherwise ->
+          mempty -- No subscription worker required
```

Comment:
> if we did try to base it on the topology, we'd then need to make some assumptions about the runtime values of those node resources (in particular, if topologyRunKademlia says True, then nrKademlia should be Just a value).

Yes this is the unfortunate truth. I use an unsafe `Maybe KademliaDHTInstance -> KademliaDHTInstance` in `src/Pos/Worker.hs` in cases where the kademlia discovery/subscription worker is needed.

It's worth noting that whether to run kademlia, and whether to run the discovery/subscription worker, are different questions. This `topologyRunKademlia` makes perfect sense and looks right to me: it determines whether `bracketKademlia` from `src/Pos/Launcher/Resources.hs` should be run. It shouldn't give a `Bool`, though, it should give a `Maybe KademliaParams`.